### PR TITLE
Kill HRGN

### LINKED ISSUE #2100: When a dialog is open, the titlebar is the wrong color
related to #1625.

image

This one's tricky - the Background of the TabRow doesn't actually change here, it's just being combined with the overlay of the ContentDialog. We'd have to compose that color ourselves in the NonClientWindow class.

### LINKED ISSUE #4744: Investigate Acrylic in titlebar / not using a `HRGN` for the titlebar
# Description of the new feature/enhancement

The title says it all.
This is another really good thing that Windows Terminal could borrow from [felixse/FluentTerminal](https://github.com/felixse/FluentTerminal).
This would basically expand the current terminal's background (including all the modifications done to it, for example a custom background image or opacity level) over to the titlebar (including the minimize, maximize and close icons). 
This would make it consistent with other Windows apps following Fluent Design (think of the Settings app and its titlebar), apart from making it look really cool.
Attached here is a screenshot of Fluent Terminal (with acrylic blur and without any background image):

Screenshot (145)

# Proposed technical implementation details (optional)

While digging for similar requests (I couldn't find an issue that addressed this specifically), I read somewhere that such a thing isn't possible given the architecture. Well, if that's the case, too bad. But I'm still opening the issue so that this remains documented somewhere separately.

### PR DESCRIPTION
**Pull request by greg904**
## Summary of the Pull Request

My understanding is that the XAML framework uses another way of getting mouse input that doesn't work with `WM_SYSCOMMAND` with `SC_MOVE`. It looks like it "steals" our mouse messages like `WM_LBUTTONDOWN`.

Before, we were cutting (with `HRGN`s) the drag bar part of the XAML islands window in order to catch mouse messages and be able to implement the drag bar that can move the window. However this "cut" doesn't only apply to input (mouse messages) but also to the graphics so we had to paint behind with the same color as the drag bar using GDI to hide the fact that we were cutting the window.

The main issue with this is that we have to replicate exactly the rendering on the XAML drag bar using GDI and this is bad because:
1. it's hard to keep track of the right color: if a dialog is open, it will cover the whole window including the drag bar with a transparent white layer and it's hard to keep track of those things.
2. we can't do acrylic with GDI

So I found another method, which is to instead put a "drag window" exactly where the drag bar is, but on top of the XAML islands window (in Z order). I've found that this makes us receive the `WM_LBUTTONDOWN` messages. That window has to have the `WS_EX_LAYERED` style and I don't know why. Also we have to recreate the window on every resize (see comments).

## PR Checklist
* [x] Closes #4744. Closes #2100.
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #4744

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
!!!!

**Comment by greg904** [comment]
It works but I wouldn't be that happy about it because it's very hacky:
1. I have to recreate a window on every resize because otherwise it mysteriously doesn't receive mouse messages  and I don't understand why.
2. I have to use `WS_EX_LAYERED` on a child window and I don't understand why.

**Comment by zadjii-msft** [comment]
I mean, I really like this. I'm certainly always hesitant about changes to our window, so I'd want to keep a careful eye on this, but just playing with this, it doesn't _seem_ like anything else has regressed.

I'm definitely curious about _why_ we need to remake a new `HWND` every time the window resizes. That's certainly really odd. Maybe the island window is placing itself on top of the draw window on a resize, and we just need to move the drag window back on top of the island window (in the z-order)? I'm no expert here so that's just a guess.

**Comment by zadjii-msft** [comment]
Would this also probably fix #4166 (by virtue of just not having a grab bar at all)?

@DHowett-MSFT and I spoke about it, and if it _doesn't_ end up working, and creates more issues that are less solvable, it's easy enough to revert before 1.0. I've played with this and I honestly don't think it's bad. @greg904 If you're okay with this going in, I think we're okay with this change.

**Comment by greg904** [comment]
It doesn't fix #4166. Now the big gray rectangle that covers the min/max controls is "invisible" when the bug happens but it's still "there" and catches mouse input:
test

> @greg904 If you're okay with this going in, I think we're okay with this change.

OK, I'm OK.

**Comment by DHowett-MSFT** [comment]
I absolutely love that this lets us hit test even over the tab bar and do, largely, whatever the *heck* we want. In the future, we could even make it ask XAML "what is underneath the cursor" before doing anything. I am so excited about this.

**Comment by greg904** [comment]
@DHowett-MSFT 

> I'd prefer we preserve the color -- it's more uniform-looking 😄

Ok. I will do this later when I have time.

> I absolutely love that this lets us hit test even over the tab bar and do, largely, whatever the heck we want. In the future, we could even make it ask XAML "what is underneath the cursor" before doing anything. I am so excited about this.

I'm not sure that this is true. Right now, either the XAML islands window receives the input either it's the "drag window" that is put on top of the drag bar.

Maybe you got that impression from the fact that I said this originally:
> I got this idea by using looking at the Settings UWP app's window using a tool called Spy++.
It looks like it does something similar to this (but not exactly this because the window's size is not exactly the same as the title bar's size) with a window called ApplicationFrameInputSinkWindow that only receive messages when I put my mouse on the drag bar:

Unfortunately I was not able to replicate the behavior from `ApplicationFrameHost.exe`. It looks like it's able to somehow control when either the `ApplicationFrameInputSinkWindow` window receives the message or when the window for XAML receives the messages. I don't understand how it works.

Another mystery is that for UWP apps, this `ApplicationFrameInputSinkWindow` window is not actually on the drag bar. It's below the drag bar and covers the whole window yet it still receives messages only when I move my cursor on the drag bar and those messages have negative Y coordinates because the cursor is not actually on the window... except it's not always the case. I have an UWP app where I called the title bar API to set everything as draggable and now this `ApplicationFrameInputSinkWindow` window doesn't even receive messages anymore when hovering over the title bar but I can still drag the window. This is very weird!

**Comment by DHowett-MSFT** [comment]
Nah, I was being more wishful/hopeful :smile: like, if we could get input on the drag bar window we could push it down into Xaml land ourselves, or some-other-how reflect it into the input queue. That's a problem for future-us.

**Comment by greg904** [comment]
> I'd prefer we preserve the color -- it's more uniform-looking 😄

done

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 1]
```diff
@@ -45,6 +45,7 @@ void IslandWindow::MakeWindow() noexcept
     wc.style = CS_HREDRAW | CS_VREDRAW;
     wc.lpfnWndProc = WndProc;
     wc.hIcon = LoadIconW(wc.hInstance, MAKEINTRESOURCEW(IDI_APPICON));
+    wc.hbrBackground = GetStockBrush(BLACK_BRUSH);
```
Comment (greg904) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp]: I need to remove this now.

**Comment by greg904** [comment]
Please note that this fixes #2100 but because I use the drag bar color to fill the background while resizing, then while resizing, if the about dialog is open the resized area will have the background color of the title bar, ignoring the fact that the dialog is open.

This is not worse than before: the bug already existed before. The bug is basically the same as #2100 but for the window background while resizing but because #2100 only mentions "the title bar" then this PR _does_ close #2100.

I don't think that this is really important because when this is fixed: https://github.com/microsoft/microsoft-ui-xaml/issues/759, we will be able to get rid of this workaround.

So I don't think it's necessary to open an issue for this small thing.

**Comment by miniksa** [comment]
@msftbot make sure @zadjii-msft and @DHowett-MSFT sign off

**Comment by zadjii-msft** [comment]
I'm tempted to let this one sit until we've got a fix for #4166, simply from the fact that it'll be easier to debug #4166 without this fix 😄 

At the same time, I also really like this fix for what doors it opens for us, so ¯\\\_(ツ)\_/¯

**Comment by greg904** [comment]
> I bet there's a way to get it working without having to recreate the
window every time.

I hope too. I have not found a way to do it unfortunately.

> I tried briefly to do the resizing thing, and yeah -- I was having
some Z order issues and couldn't even get the window to reliably paint
(i tried making it red so that i could see what was happening)
Weirdly, adding a WM_PAINT handler made the window totally
hit-testable even when it was resized... perhaps there's something
interesting here?

I have tried adding a WM_PAINT handler in the drag window that does:
```cpp
case WM_PAINT:
{
    PAINTSTRUCT ps;
    BeginPaint(window, &ps);
    EndPaint(window, &ps);
    return 0;
}
```

And using `MoveWindow` to move the drag window in `_RecreateDragBarWindow` instead of recreating.

But it doesn't work.

> The drag region is still hit testable when the application has been
fullscreened; maybe we should suppress it?

Fixed in c3759587.

> It doesn't feel incorrect; the window can be interacted with just like
normal.

Yes.

> Should we be trying to get the WM_NC events inside this window
instead of tunneling normal events through to the parent window as NC
events? I'm not sure on this one at all. 😄

I think that in the drag window proc, either:
* We tunnel the normal events through to the parent window as NC
events so that `DefWindowProc` will be called with the window parent HWND and will handle the dragging/resizing for us.
* We could call `DefWindowProc` directly with the window parent HWND instead of `SendMessage`.
* We could try to handle resizing/dragging ourselves by sending WM_SYSCOMMAND with SC_MOVE/SC_SIZE. It's probably what `DefWindowProc` already does.

But I think that in all cases, we have to tunnel something to the parent window: initially, I tried to send WM_SYSCOMMAND events with SC_MOVE to the drag bar window and what happened was that when I clicked on the drag bar, it would drag the drag bar instead of the whole window. So I made it send the message to the parent window instead. I'm no expert at this.

**Comment by greg904** [comment]
Wow so now (maybe before too?) the drag bar is broken in Release mode but works in Debug. This is bad. I don't know why.

**Comment by greg904** [comment]
BTW I'm not sure how to use the spell checker bot. I ran the command to accept the changes but it created the `.github/actions/spell-check/whitelist/c3759587195744a3a2e01e2be4ee44ef1f6ab100.txt` file and I don't see any file with the same weird name in that directory so I don't really understand.

**Comment by jsoref** [comment]
@greg904 The filename it suggests is not specific to this project (it's based on a git sha) -- When there's only one file, it just suggests that file, but when there are multiple, it doesn't really know which to pick.

It would be better if you picked one of the existing files. A future run of the checker (when you merge again) would include a bit better advice. You could put your changes into the whitelist file, or maybe the api file (when it arrives). Or wherever a reviewer suggests.

**Comment by DHowett-MSFT** [comment]
@greg904 I love this pull request so much that I pulled it into a testing branch over here on our repo and made a couple changes. I got it working without recreating the window, in both debug and release mode. I'm going to submit a pull request from that branch and you will, of course, get full credit on the merge. :)

**Comment by DHowett** [comment]
#5485

**Comment by greg904** [comment]
Ok so I should probably close this PR then. It's cool that you found a way to make it work. When I saw that it didn't work in Release mode I was mad 🤣, I thought it was going to be impossible to debug because I don't even understand how some parts of the code I wrote work.

**Pull request closed without merging**